src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -14,592 +14,558 @@ namespace Microsoft.CodeAnalysis.CSharp
-                                    unwrappedArgsBuilder.Add(singleArgument);
-                                    break;
-                                // This complicated check is for a sequence, which wraps a span around single char.
-                                // The sequence needs to have this shape: `{ locals: <none>, sideEffects: temp = <original char expression>, result: new ReadOnlySpan<char>(in temp) }`
-                                case BoundSequence
-                                {
-                                    Locals.Length: 0,
-                                    SideEffects: [BoundAssignmentOperator { Right.Type.SpecialType: SpecialType.System_Char } assignment],
-                                    Value: BoundObjectCreationExpression { Constructor: var objectCreationConstructor, Arguments: [BoundLocal constructorLocal] }
-                                } when constructorLocal == assignment.Left &&
-                                       locals.Remove(constructorLocal.LocalSymbol) &&
-                                       (object)objectCreationConstructor.OriginalDefinition == _compilation.GetSpecialTypeMember(SpecialMember.System_ReadOnlySpan_T__ctor_Reference) &&
-                                       objectCreationConstructor.ContainingType.IsReadOnlySpanChar():
-                                    var wrappedExpr = ConvertConcatExprToString(assignment.Right);
-                                    unwrappedArgsBuilder.Add(wrappedExpr);
-                                    break;
-                                default:
-                                    locals.Free();
-                                    unwrappedArgsBuilder.Free();
-                                    arguments = default;
-                                    return false;
-                            }
-                        }
-
-                        if (locals.Count > 0)
-                        {
-                            // Not all locals are part of a known shape
-                            locals.Free();
-                            unwrappedArgsBuilder.Free();
-                            arguments = default;
-                            return false;
-                        }
+                        binaryOperator = left;
+                    }
+                    else
+                    {
+                        self.VisitAndAddConcatArgumentInReverseOrder(binaryOperator.Left, argumentAlreadyVisited: false, arguments, ref concatMethods);
+                        break;
+                    }
+                }
 
-                        locals.Free();
-                        arguments = unwrappedArgsBuilder.ToImmutableAndFree();
+                static bool shouldRecurse(BoundExpression expr, [NotNullWhen(true)] out BoundBinaryOperator? binaryOperator)
```

Comment:
nit: I think we try to avoid nested local functions. Consider popping this one out

Reply:
I prefer to keep local functions as scoped as possible. The outer functions/methods do not need access to this, so I will keep it where it is.